kie-server-parent/kie-server-services/kie-server-services-optaplanner/src/main/java/org/kie/server/services/optaplanner/OptaplannerKieServerExtension.java

```diff
@@ -0,0 +1,140 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.services.optaplanner;
+
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.services.api.*;
+import org.kie.server.services.impl.KieServerImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.*;
+
+public class OptaplannerKieServerExtension
+        implements KieServerExtension {
+
+    private static final Logger logger = LoggerFactory.getLogger( OptaplannerKieServerExtension.class );
+
+    public static final String EXTENSION_NAME = "Optaplanner";
+
+    private static final Boolean droolsDisabled = Boolean.parseBoolean(System.getProperty(KieServerConstants.KIE_DROOLS_SERVER_EXT_DISABLED, "false"));
+    private static final Boolean disabled = Boolean.parseBoolean( System.getProperty( KieServerConstants.KIE_OPTAPLANNER_SERVER_EXT_DISABLED, "false" ) );
+
+    private KieServerRegistry registry;
+    private SolverServiceBase solverServiceBase;
+
+    // Optaplanner requires threads to run solvers
+    // asynchronously. We can't use JEE managed threads
+    // for this because they would red-flag the threads
+    // that run for long periods. This first implementation
+    // will use a standard java thread pool for the job.
+    // If necessary, we will need to look for alternatives
+    // in the future.
+    private ExecutorService threadPool = null;
+
+    private List<Object> services = new ArrayList<Object>();
+
+    @Override
+    public boolean isActive() {
+        return  disabled == false && droolsDisabled == false;
+    }
+
+    @Override
+    public void init(KieServerImpl kieServer, KieServerRegistry registry) {
+        KieServerExtension droolsExtension = registry.getServerExtension( "Drools" );
+        if ( droolsExtension == null ) {
+            logger.warn( "No Drools extension available, quiting..." );
+            return;
+        }
+        this.registry = registry;
+        // the following threadpool will have a max thread count equal to the number of cores on the machine.
+        // if new jobs are submited and all threads are busy, the reject policy will kick in.
+        this.threadPool = new ThreadPoolExecutor(2, // core size
+                                                 Runtime.getRuntime().availableProcessors(), // max size
+                                                 120, // idle timeout
+                                                 TimeUnit.SECONDS,
+                                                 new ArrayBlockingQueue<Runnable>(Runtime.getRuntime().availableProcessors())); // queue with a size
```

Comment:
From experience, I 'd recommend to take a more conservative number for the threadpool size:

    threadPoolSize = (availableProcessors < 4 ? availableProcessors : availableProcessors - 1);

Why? If you have 8 processors and 8 solver threads running in parallel, which each take 100% of their processor (remember that they don't do IO, so they never block), then any REST request coming in will need to rely on the JVM's round-robin thread scheduling to park at least 1 of the running solver threads to get some time to be handled (and for the OS and for EAP threads). I believe this will affect response time (= latency). If we have 4 or more processors, I think the tradeoff is worth to reserve 1 processor to optimize latency over throughput (note that such a reservation is virtual because solver threads jump to another cpu all the time, but you will see that total cpu time isn't maxed out).

In benchmarks, I am even more conservative, only using half of the processorSize by default. This is mainly because on hardware with hyper-threading, stressing more than half of the CPU's clearly reduces the average speed per CPU.

In any case: in the long run that thread pool size will need to be configurable for the SysOps (through JMX I suppose).